server/src/main/java/org/opensearch/common/cache/stats/StatsHolder.java

```diff
@@ -0,0 +1,200 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.cache.stats;
+
+import java.util.List;
+import java.util.function.Consumer;
+
+import static org.opensearch.common.cache.stats.MultiDimensionCacheStats.MDCSDimensionNode;
+
+/**
+ * A class caches use to internally keep track of their stats across multiple dimensions.
+ * Not intended to be exposed outside the cache; for this, use statsHolder.getCacheStats() to create an immutable
+ * copy of the current state of the stats.
+ *
+ * @opensearch.experimental
+ */
+public class StatsHolder {
+
+    // The list of permitted dimensions. Should be ordered from "outermost" to "innermost", as you would like to
+    // aggregate them in an API response.
+    private final List<String> dimensionNames;
+
+    // A tree structure based on dimension values, which stores stats values in its leaf nodes.
+    // Non-leaf nodes have stats matching the sum of their children.
+    private final DimensionNode statsRoot;
+
+    public StatsHolder(List<String> dimensionNames) {
+        this.dimensionNames = dimensionNames;
+        this.statsRoot = new DimensionNode(null); // The root node has no dimension value associated with it, only children
+        statsRoot.createChildrenMap();
```

Comment:
Why do we premptively need to initialize the childrenMap for statsRoot?

Reply:
You're right, we shouldn't have to.